src/main/java/heronarts/lx/dmx/DmxGridModulator.java

```diff
@@ -0,0 +1,117 @@
+/**
+ * Copyright 2023- Justin K. Belcher, Mark C. Slee, Heron Arts LLC
+ *
+ * This file is part of the LX Studio software library. By using
+ * LX, you agree to the terms of the LX Studio Software License
+ * and Distribution Agreement, available at: http://lx.studio/license
+ *
+ * Please note that the LX license is not open-source. The license
+ * allows for free, non-commercial use.
+ *
+ * HERON ARTS MAKES NO WARRANTY, EXPRESS, IMPLIED, STATUTORY, OR
+ * OTHERWISE, AND SPECIFICALLY DISCLAIMS ANY WARRANTY OF
+ * MERCHANTABILITY, NON-INFRINGEMENT, OR FITNESS FOR A PARTICULAR
+ * PURPOSE, WITH RESPECT TO THE SOFTWARE.
+ *
+ * @author Mark C. Slee <mark@heronarts.com>
+ * @author Justin K. Belcher <justin@jkb.studio>
+ */
+
+package heronarts.lx.dmx;
+
+import heronarts.lx.LXCategory;
+import heronarts.lx.modulator.LXModulator;
+import heronarts.lx.parameter.DiscreteParameter;
+import heronarts.lx.parameter.LXParameter;
+
+/**
+ * Maps a sequential set of DMX channels onto a 2D grid of size [rows x columns]
+ */
+@LXModulator.Global("DMX Grid")
+@LXModulator.Device("DMX Grid")
+@LXCategory(LXCategory.DMX)
+public class DmxGridModulator extends DmxModulator {
```

Comment:
I'm not sure if this should really be called a Grid rather than just a generic range. The fact that there are rows and columns is not really relevant to its operation... the only thing the algorithm cares about is total number of values being averaged.

There could still be a helper for row/col based indexing, but that's really just a utility helper-method. The function of the modulator itself is to average a range.

I think "Range" as a name would probably intuitively correspond to this behavior. Or if we have a DMX section, than really I think this could even just be called "Average" - that's actually the clearest really.

Is there a particular use case you guys have where this being a grid is really relevant?

Reply:
Ah sorry this one wasn't clear.  The purpose is not to extract an average... I was thinking modulators internally contain a single value and the average was more useful than zero.  The main purpose here was to expose the getValue(row, column) method.

We're using it to receive a low-resolution grid from a grandMA console, overlayed onto the fixtures of the car. Allowing the console to map a swipe effect across their fixtures outside of the car and within the car itself.  Demo vid:
https://www.youtube.com/watch?v=sPlDItBaxd0

Reply:
The use case is cool but I'm not really seeing why there is a modulator here. Shouldn't these just be parameters and a helper method on your Dmx Grid pattern implementation? Is your Dmx Grid pattern looking for a global modulator of the DmxGrid type? That feels hacky/messy.

This one feels like it ought to just live in pattern land as utility code, not as a user-facing modulator. The grid settings in the modulator don't do anything for a user, only for a programmer that is coding against the helper method API. And I don't think you need all this overhead of the modulator separately caching its own `values[][]` array and recomputing it on every draw loop. Feels like your pattern can just ask the DmxEngine directly for the values it wants when it needs them, and there's no extra processing that way.

Reply:
Yep, tunnel vision on routing everything through modulators.  Removed.